File: src/main/java/io/github/daviddenton/metrique/metrics/ChildMetricsClient.java
package io.github.daviddenton.metrique.metrics;

public class ChildMetricsClient implements MetricsClient {

    private final MetricsClient delegate;
    private final String prefix;

    public ChildMetricsClient(MetricsClient delegate, String prefix) {
        this.delegate = delegate;
        this.prefix = prefix;
    }

    private MetricName childMetric(MetricName name) {
        return new MetricName(prefix, name.toString());
    }

    public void decrement(MetricName name) {
        delegate.decrement(childMetric(name));
    }

    public void count(MetricName name, Long value) {
        delegate.count(childMetric(name), value);
    }

    public void increment(MetricName name) {
        delegate.increment(childMetric(name));
    }

    public void gauge(MetricName name, Long value) {
        delegate.gauge(childMetric(name), value);
    }

    public void time(MetricName name, Long value) {
        delegate.time(childMetric(name), value);
    }
}

File: src/main/java/io/github/daviddenton/metrique/metrics/ChildMetricsClient.scala
package io.github.daviddenton.metrique.metrics

class ChildMetricsClient(delegate: MetricsClient, prefix: String) extends MetricsClient {

  private def childMetric(name: MetricName): MetricName = {
    MetricName(prefix + "." + name)
  }

  override def decrement(name: MetricName): Unit = delegate.decrement(childMetric(name))

  override def count(name: MetricName, value: Long): Unit = delegate.count(childMetric(name), value)

  override def increment(name: MetricName): Unit = delegate.increment(childMetric(name))

  override def gauge(name: MetricName, value: Long): Unit = delegate.gauge(childMetric(name), value)

  override def time(name: MetricName, value: Long): Unit = delegate.time(childMetric(name), value)
}
